### PR TITLE
Update VerifyMerkleBranch to match ETH2.0 spec

### LINKED ISSUE #6705: `is_valid_merkle_branch` is out of date
Our `is_valid_merkle_branch` implementation:
```go
func VerifyMerkleBranch(root []byte, item []byte, merkleIndex int, proof [][]byte) bool {
	node := bytesutil.ToBytes32(item)
	currentIndex := merkleIndex
	for i := 0; i < len(proof); i++ {
		if currentIndex%2 != 0 {
			node = hashutil.Hash(append(proof[i], node[:]...))
		} else {
			node = hashutil.Hash(append(node[:], proof[i]...))
		}
		currentIndex = currentIndex / 2
	}
	return bytes.Equal(root, node[:])
}
```

Spec's `is_valid_merkle_branch` implementation:
```go
def is_valid_merkle_branch(leaf: Bytes32, branch: Sequence[Bytes32], depth: uint64, index: uint64, root: Root) -> bool:
    """
    Check if ``leaf`` at ``index`` verifies against the Merkle ``root`` and ``branch``.
    """
    value = leaf
    for i in range(depth):
        if index // (2**i) % 2:
            value = hash(branch[i] + value)
        else:
            value = hash(value + branch[i])
    return value == root
```

Spec uses input `depth` to loop through each layer while our implementation relies on length of the proofs. One can argue it's the same but to be safe we should update to the spec version. This also makes it more extensible to be used for phase 1

### PR DESCRIPTION
**Pull request by fabdarice**
**What type of PR is this?**
> Other

**What does this PR do? Why is it needed?**
Update `VerifyMerkleBranch function to match the ETh2.0 spec.

**Which issues(s) does this PR fix?**

Fixes #6705

**Other notes for review**

### CODE REVIEW

File: shared/trieutil/sparse_merkle.go [hunk 1]
```diff
@@ -183,15 +185,14 @@ func (m *SparseMerkleTrie) ToProto() *protodb.SparseMerkleTrie {
...
-		if currentIndex%2 != 0 {
+	for i := 0; i < int(params.BeaconConfig().DepositContractTreeDepth); i++ {
+		if (uint64(merkleIndex) / mathutil.PowerOf2(uint64(i)) % 2) != 0 {
 			node = hashutil.Hash(append(proof[i], node[:]...))
 		} else {
 			node = hashutil.Hash(append(node[:], proof[i]...))
 		}
-		currentIndex = currentIndex / 2
 	}
+	node = hashutil.Hash(append(node[:], proof[int(params.BeaconConfig().DepositContractTreeDepth)]...))
```
Comment (fabdarice) [inline shared/trieutil/sparse_merkle.go:195]: Had to add this part, looking at how Root() function is calculation the Merkle root.

Would someone kind enough explain what is the purpose of `enc` below?:
```
func (m *SparseMerkleTrie) Root() [32]byte {
	enc := [32]byte{}
	binary.LittleEndian.PutUint64(enc[:], uint64(len(m.originalItems)))
	return hashutil.Hash(append(m.branches[len(m.branches)-1][0], enc[:]...))
}
```

File: shared/trieutil/sparse_merkle.go [hunk 2]
```diff
@@ -183,15 +185,14 @@ func (m *SparseMerkleTrie) ToProto() *protodb.SparseMerkleTrie {
 // VerifyMerkleBranch verifies a Merkle branch against a root of a trie.
 func VerifyMerkleBranch(root []byte, item []byte, merkleIndex int, proof [][]byte) bool {
 	node := bytesutil.ToBytes32(item)
-	currentIndex := merkleIndex
-	for i := 0; i < len(proof); i++ {
-		if currentIndex%2 != 0 {
+	for i := 0; i < int(params.BeaconConfig().DepositContractTreeDepth); i++ {
```
Comment (terencechain) [inline shared/trieutil/sparse_merkle.go:188]: `params.BeaconConfig().DepositContractTreeDepth` should be a parameter

See `depth: uint64` in the spec definition

File: shared/trieutil/sparse_merkle_test.go [hunk 3]
```diff
@@ -150,7 +150,7 @@ func TestMerkleTrie_VerifyMerkleProof_TrieUpdated(t *testing.T) {
 		{3},
 		{4},
 	}
-	m, err := GenerateTrieFromItems(items, int(params.BeaconConfig().DepositContractTreeDepth)+1)
```
Comment (prestonvanloon) [inline shared/trieutil/sparse_merkle_test.go:153]: Why change this?

Comment (fabdarice) [reply shared/trieutil/sparse_merkle_test.go:153]: probably tie back to @terencechain comment.
As I wasn't passing the depth of the trie as an input of `VerifyMerkleBranch`, the verification always assume a depth of `int(params.BeaconConfig().DepositContractTreeDepth)`.

This test was creating a Trie of depth +1, so I updated it.

Will update the implementation to pass depth as an input per @terencechain comment above.

**Comment by fabdarice** [comment]
resuming work on this tomorrow. (sorry been busy at work).
A few unit tests are failing.

**Comment by rauljordan** [comment]
Hi @fabdarice, will you be able to pick up work again on this PR soon? We want to avoid it going stale

**Comment by fab-spring** [comment]
> Hi @fabdarice, will you be able to pick up work again on this PR soon? We want to avoid it going stale

Absolutely, I'll tackle it tonight.
Apologies, work has picked up but this is 90% done.

**Comment by rauljordan** [comment]
PTAL @terencechain

**Review by terencechain — DISMISSED** [review]

**Pull request merged**